README.md
Welcome to _TRUSTplus Wallet_, a standalone TRUSTplus payment app for your Android device!

This project contains several sub-projects:

 * __wallet__:
     The Android app itself. This is probably what you're searching for.
 * __market__:
     App description and promo material for the Google Play app store.
 * __integration-android__:
     A tiny library for integrating digitial payments into your own Android app
     (e.g. donations, in-app purchases).
 * __sample-integration-android__:
     A minimal example app to demonstrate integration of digital payments into
     your Android app.

You can build all sub-projects at once using Maven:

`mvn clean install -DskipTests`


wallet/CHANGES
TRUSTplus Wallet
---------------
v0.1.0
* First version

Darkcoin Wallet
---------------
v1.0.1
* Fix bug downloading headers
* updated checkpoints

v1.0.0
* Added DNS Seeds

v0.10.2
* updated for RC2.

v0.1.1
* updated all icons for Darkcoin

v0.1.0
* First version
digitalcoin Wallet
-----------------
v0.1.5-beta
* update to bitcoin-wallet 3.46
* added Litecoin's PeerDBDiscovery System
* added IRC bootstrap for more peers
* Send Coins shows the transaction fee (currently the minimum fee)

v0.1.4-beta
* permanent fix for bugs relating to pause activities (removed temp fixes and support-v4r13)

v0.1.3-beta
* fixed several bugs relating to pausing activities

v0.1.2-beta
* fixed checkpoints to work for the Version 1.0 blocks (between 476280 and 523800)
* added two nodes from a list of addnodes from the bitcointalk topic

v0.1.1-beta
* renamed to hashengineering.digitalcoin.wallet
* added support for importing Bitcoin Private Keys from first digitalcoin app (Crypto-Expert)

v1.0.1.5
* Fixed menu item doubling on rotate in address screen - langerhans (https://github.com/langerhans/dogecoin-wallet-new)
* replaced block explorer with blockr.io instead of the apparently dead dgc.cryptocoinexplorer.com
* Added BTER as a backup for getting exchange rates
* added mBTC as an exchange rate.
* added Disconnect feature back in
* updated to bitcoin-wallet v3.39

v1.0.1.4
* Updated to v3.31 (bitcoin-wallet)
* Few bug fixes (import parser)

v1.0.1.3

* Added checkpoints file for faster blockchain sync
* Sync with bitcoin-wallet 3.30

v1.0.1.2

* Fix broken diff adjust in 1.0.1.1

v1.0.1.1

*  updated digitalcoin 1.0.1 - difficulty adjustment fork

v1.00
* Forked bitcoin-wallet v3.23 for digitalcoin V1.00 (http://digitalcoin.co)

Bitcoin Wallet
-------------------
v3.46

* New translation into Slovenian.
* Update Hebrew translation.
* Fix missing transactions due to once dusty outputs which are standard as of Bitcoin Core 0.9.
* Preparation for fetching payment requests via Bluetooth.
* Based on bitcoinj 0.11.2.

v3.44-v3.45

* Due to the OpenSSL "Heartbleed" vulnerability, removed all HTTPS network access on Android 4.1.1.
  This means exchange rates will not be available, and the payment protocol cannot fetch
  signature and cannot pay directly via HTTPS. The chances of an actual exploitability on
  Android are very slim, but we want to stay on the safe side until things are clearer.
  Also see http://heartbleed.com/

v3.41-v3.43

* Small bugfixes.

v3.40

* Support for µBTC denomination, for people who are used to low valued currencies.
* Always use BIP70 when sending payments directly via Bluetooth.
* Small bugfixes.

v3.39

* Usage of BIP70 when initiating payments via NFC (tap-to-pay) is now a supported feature.
* New translation into Arabic.

v3.38

* Supports BIP70 payment protocol for in-app payments.
* Accepting direct payments via Bluetooth is now a supported feature.

v3.37

* New translation into Japanese.

v3.36

* Small bugfix.

v3.35

* Protect wallet against dust spam.
* Fixed some bugs in the payment protocol.
* Based on bitcoinj 0.11.1.

v3.34

* Small bugfix.

v3.33

* Small bugfixes.

v3.32

* Support for BIP72 URI extension for payment protocol.

v3.31

* Default denomination has changed to millibitcoin (mBTC).
* Cross-calculated exchange rates for currencies without Bitcoin volume. Thanks to BitcoinAverage.
* Better handling of errors when sending coins, specifically when funds are exceeded.
* Labs: Uses BIP70 payment protocol for tap-to-pay and scan-to-pay, if enabled in the settings.
* New translations into Traditional Chinese, Macedonian and Esperanto. Updated Spanish.
* Resolved issues with wallets with many transactions.
* Based on bitcoinj 0.11.

v3.30

* improved battery usage
* new Danish translation
* updated Polish translation

v3.29

* fixed issues with fetching exchange rates
* new Slovak translation
* updated French, Hungarian, Italian and Portuguese translations

v3.28

* for those who still use the unsecure address (pre-August), moves coins to secure addresses
* updated Russian and Turkish translations

v3.27

* small improvements for future bugfixing

v3.26

* improve battery usage
* small bugfixes

v3.25

* support for mBTC denomination
* also use weighted exchange rates from BitcoinAverage
* fixed issue with camera autofocus on some Samsung Galaxy devices

v3.24

* fixed issue with stuck outgoing transactions

>>>>>>> upstream/prod
v3.23

* fixed issue with detecting online state

v3.22

* sound effect when sending a payment
* fixed rare issues with transactions missing or shown as unconfirmed
* Hungarian translation
* based on bitcoinj 0.10.2

v3.21

* bugfix

v3.20

* small improvements and bugfixes

v3.19

* use findbugs
* small bugfixes

v3.18

* Indonesian translation
* bluetooth offline transactions
* revamp of qr-code and nfc based offline transactions
* remove transaction details screen, link to blockexplorer.com instead

v3.17

* fixed issue with emptying wallet
* based on bitcoinj 0.10.1

v3.16

* backup reminder when creating new keys
* don't trigger blockchain sync because of plugging power

v3.15

* auto-rotate weak keys
* option for easy emptying wallet
* based on bitcoinj 0.10

v3.14

* Vietnamese and Portuguese (Brazil) translations
* small improvements

v3.13

* bugfix

v3.12

* direct button on main screen for scanning a Bitcoin request QR code
* extended logging

v3.11

* some small improvements and bugfixes

v3.10

* embed currency calculator into send and request coins forms
* change default Bitcoin precision from 8 to 4

v3.09

* calculate correct fee to ensure speedy blockchain confirmation
* based on bitcoinj 0.9

v3.08

* bugfixes

v3.07

* make request and send coins action buttons more visible
* stay connected a little bit longer if there was transaction activity on the wallet

v3.06

* merge peer monitor and block explorer into network monitor
* Hebrew translation

v3.05

* backup reminder for first-time users

v3.01-v3.04


wallet/src/de/schildbach/wallet/util/Crypto.java
/*
 * Copyright 2012-2014 the original author or authors.
 *
 * Licensed under the MIT license (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *    http://opensource.org/licenses/mit-license.php
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package de.schildbach.wallet.util;

import android.util.Base64;
import de.schildbach.wallet.Constants;
import org.spongycastle.crypto.*;
import org.spongycastle.crypto.engines.AESFastEngine;
import org.spongycastle.crypto.generators.OpenSSLPBEParametersGenerator;
import org.spongycastle.crypto.modes.CBCBlockCipher;
import org.spongycastle.crypto.paddings.PaddedBufferedBlockCipher;
import org.spongycastle.crypto.params.ParametersWithIV;

import javax.annotation.Nonnull;
import java.io.*;
import java.security.SecureRandom;
import java.util.Arrays;

/**
 * This class encrypts and decrypts a string in a manner that is compatible with OpenSSL.
 * 
 * If you encrypt a string with this class you can decrypt it with the OpenSSL command: openssl enc -d -aes-256-cbc -a
 * -in cipher.txt -out plain.txt -pass pass:aTestPassword
 * 
 * where: cipher.txt = file containing the cipher text plain.txt - where you want the plaintext to be saved
 * 
 * substitute your password for "aTestPassword" or remove the "-pass" parameter to be prompted.
 * 
 * @author jim
 * @author Andreas Schildbach
 */
public class Crypto
{
	/**
	 * number of times the password & salt are hashed during key creation.
	 */
	private static final int NUMBER_OF_ITERATIONS = 1024;

	/**
	 * Key length.
	 */
	private static final int KEY_LENGTH = 256;

	/**
	 * Initialization vector length.
	 */
	private static final int IV_LENGTH = 128;

	/**
	 * The length of the salt.
	 */
	private static final int SALT_LENGTH = 8;

	/**
	 * OpenSSL salted prefix text.
	 */
	private static final String OPENSSL_SALTED_TEXT = "Salted__";

	/**
	 * OpenSSL salted prefix bytes - also used as magic number for encrypted key file.
	 */
	private static final byte[] OPENSSL_SALTED_BYTES = OPENSSL_SALTED_TEXT.getBytes(Constants.UTF_8);

	/**
	 * Magic text that appears at the beginning of every OpenSSL encrypted file. Used in identifying encrypted key
	 * files.
	 */
	private static final String OPENSSL_MAGIC_TEXT = new String(encodeBase64(Crypto.OPENSSL_SALTED_BYTES), Constants.UTF_8).substring(0,
			Crypto.NUMBER_OF_CHARACTERS_TO_MATCH_IN_OPENSSL_MAGIC_TEXT);

	private static final int NUMBER_OF_CHARACTERS_TO_MATCH_IN_OPENSSL_MAGIC_TEXT = 10;

	private static final SecureRandom secureRandom = new SecureRandom();

	/**
	 * Get password and generate key and iv.
	 * 
	 * @param password
	 *            The password to use in key generation
	 * @param salt
	 *            The salt to use in key generation
	 * @return The CipherParameters containing the created key
	 */
	private static CipherParameters getAESPasswordKey(final char[] password, final byte[] salt)
	{
		final PBEParametersGenerator generator = new OpenSSLPBEParametersGenerator();
		generator.init(PBEParametersGenerator.PKCS5PasswordToBytes(password), salt, NUMBER_OF_ITERATIONS);

		final ParametersWithIV key = (ParametersWithIV) generator.generateDerivedParameters(KEY_LENGTH, IV_LENGTH);

		return key;
	}

	/**
	 * Password based encryption using AES - CBC 256 bits.
	 * 
	 * @param plainText
	 *            The text to encrypt
	 * @param password
	 *            The password to use for encryption
	 * @return The encrypted string
	 * @throws IOException
	 */
	public static String encrypt(@Nonnull final String plainText, @Nonnull final char[] password) throws IOException
	{
		final byte[] plainTextAsBytes = plainText.getBytes(Constants.UTF_8);

		final byte[] encryptedBytes = encrypt(plainTextAsBytes, password);

		// OpenSSL prefixes the salt bytes + encryptedBytes with Salted___ and then base64 encodes it
		final byte[] encryptedBytesPlusSaltedText = concat(OPENSSL_SALTED_BYTES, encryptedBytes);

		return new String(encodeBase64(encryptedBytesPlusSaltedText), Constants.UTF_8);
	}

	/**
	 * Password based encryption using AES - CBC 256 bits.
	 * 
	 * @param plainBytes
	 *            The bytes to encrypt
	 * @param password
	 *            The password to use for encryption
	 * @return SALT_LENGTH bytes of salt followed by the encrypted bytes.
	 * @throws IOException
	 */
	private static byte[] encrypt(final byte[] plainTextAsBytes, final char[] password) throws IOException
	{
		try
		{
			// Generate salt - each encryption call has a different salt.
			final byte[] salt = new byte[SALT_LENGTH];
			secureRandom.nextBytes(salt);

			final ParametersWithIV key = (ParametersWithIV) getAESPasswordKey(password, salt);

			// The following code uses an AES cipher to encrypt the message.
			final BufferedBlockCipher cipher = new PaddedBufferedBlockCipher(new CBCBlockCipher(new AESFastEngine()));
			cipher.init(true, key);
			final byte[] encryptedBytes = new byte[cipher.getOutputSize(plainTextAsBytes.length)];
			final int processLen = cipher.processBytes(plainTextAsBytes, 0, plainTextAsBytes.length, encryptedBytes, 0);
			final int doFinalLen = cipher.doFinal(encryptedBytes, processLen);

			// The result bytes are the SALT_LENGTH bytes followed by the encrypted bytes.
			return concat(salt, Arrays.copyOf(encryptedBytes, processLen + doFinalLen));
		}
		catch (final InvalidCipherTextException x)
		{
			throw new IOException("Could not encrypt bytes", x);
		}
		catch (final DataLengthException x)
		{
			throw new IOException("Could not encrypt bytes", x);
		}
	}

	/**
	 * Decrypt text previously encrypted with this class.
	 * 
	 * @param textToDecode
	 *            The code to decrypt
	 * @param password
	 *            password to use for decryption
	 * @return The decrypted text
	 * @throws IOException
	 */
	public static String decrypt(@Nonnull final String textToDecode, @Nonnull final char[] password) throws IOException
	{
		final byte[] decodeTextAsBytes = decodeBase64(textToDecode.getBytes(Constants.UTF_8));

		if (decodeTextAsBytes.length < OPENSSL_SALTED_BYTES.length)
			throw new IOException("out of salt");

		final byte[] cipherBytes = new byte[decodeTextAsBytes.length - OPENSSL_SALTED_BYTES.length];
		System.arraycopy(decodeTextAsBytes, OPENSSL_SALTED_BYTES.length, cipherBytes, 0, decodeTextAsBytes.length - OPENSSL_SALTED_BYTES.length);

		final byte[] decryptedBytes = decrypt(cipherBytes, password);

		return new String(decryptedBytes, Constants.UTF_8).trim();
	}

	/**
	 * Decrypt bytes previously encrypted with this class.
	 * 
	 * @param bytesToDecode
	 *            The bytes to decrypt
	 * @param passwordbThe
	 *            password to use for decryption
	 * @return The decrypted bytes
	 * @throws IOException
	 */
	private static byte[] decrypt(final byte[] bytesToDecode, final char[] password) throws IOException
	{
		try
		{
			// separate the salt and bytes to decrypt
			final byte[] salt = new byte[SALT_LENGTH];

			System.arraycopy(bytesToDecode, 0, salt, 0, SALT_LENGTH);

			final byte[] cipherBytes = new byte[bytesToDecode.length - SALT_LENGTH];
			System.arraycopy(bytesToDecode, SALT_LENGTH, cipherBytes, 0, bytesToDecode.length - SALT_LENGTH);

			final ParametersWithIV key = (ParametersWithIV) getAESPasswordKey(password, salt);

			// decrypt the message
			final BufferedBlockCipher cipher = new PaddedBufferedBlockCipher(new CBCBlockCipher(new AESFastEngine()));
			cipher.init(false, key);

			final byte[] decryptedBytes = new byte[cipher.getOutputSize(cipherBytes.length)];
			final int processLen = cipher.processBytes(cipherBytes, 0, cipherBytes.length, decryptedBytes, 0);
			final int doFinalLen = cipher.doFinal(decryptedBytes, processLen);

			return Arrays.copyOf(decryptedBytes, processLen + doFinalLen);
		}
		catch (final InvalidCipherTextException x)
		{
			throw new IOException("Could not decrypt input string", x);
		}
		catch (final DataLengthException x)
		{
			throw new IOException("Could not decrypt input string", x);
		}
	}

	private static byte[] encodeBase64(byte[] decoded)
	{
		return Base64.encode(decoded, Base64.DEFAULT);
	}

	private static byte[] decodeBase64(byte[] encoded) throws IOException
	{
		try
		{
			return Base64.decode(encoded, Base64.DEFAULT);
		}
		catch (final IllegalArgumentException x)
		{
			throw new IOException("illegal base64 padding", x);
		}
	}

	/**
	 * Concatenate two byte arrays.
	 */
	private static byte[] concat(final byte[] arrayA, final byte[] arrayB)
	{
		final byte[] result = new byte[arrayA.length + arrayB.length];
		System.arraycopy(arrayA, 0, result, 0, arrayA.length);
		System.arraycopy(arrayB, 0, result, arrayA.length, arrayB.length);

		return result;
	}

	public final static FileFilter OPENSSL_FILE_FILTER = new FileFilter()
	{
		private final char[] buf = new char[OPENSSL_MAGIC_TEXT.length()];

		@Override
		public boolean accept(final File file)
		{
			Reader in = null;
			try
			{
				in = new InputStreamReader(new FileInputStream(file), Constants.UTF_8);
				if (in.read(buf) == -1)
					return false;
				final String str = new String(buf);
				if (!str.toString().equals(OPENSSL_MAGIC_TEXT))
					return false;
				return true;
			}
			catch (final IOException x)
			{
				return false;
			}
			finally
			{
				if (in != null)
				{
					try
					{
						in.close();
					}
					catch (final IOException x2)
					{
					}
				}
			}
		}
	};
}


wallet/src/de/schildbach/wallet/util/PaymentProtocol.java
/*
 * Copyright 2014 the original author or authors.
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

package de.schildbach.wallet.util;

import java.math.BigInteger;
import java.util.ArrayList;
import java.util.List;

import javax.annotation.Nonnull;
import javax.annotation.Nullable;

import com.google.bitcoin.core.CoinDefinition;
import org.bitcoin.protocols.payments.Protos;

import com.google.bitcoin.core.Address;
import com.google.bitcoin.core.ScriptException;
import com.google.bitcoin.core.Transaction;
import com.google.bitcoin.protocols.payments.PaymentRequestException;
import com.google.bitcoin.protocols.payments.PaymentSession;
import com.google.bitcoin.protocols.payments.PaymentSession.PkiVerificationData;
import com.google.bitcoin.script.Script;
import com.google.bitcoin.script.ScriptBuilder;
import com.google.protobuf.ByteString;
import com.google.protobuf.InvalidProtocolBufferException;
import com.google.protobuf.UninitializedMessageException;

import de.schildbach.wallet.Constants;
import de.schildbach.wallet.PaymentIntent;

/**
 * @author Andreas Schildbach
 */
public final class PaymentProtocol
{
	public static final String MIMETYPE_PAYMENTREQUEST = "application/"+ CoinDefinition.coinName+"-paymentrequest"; // BIP 71
	public static final String MIMETYPE_PAYMENT = "application/"+ CoinDefinition.coinName+"-payment"; // BIP 71
	public static final String MIMETYPE_PAYMENTACK = "application/"+ CoinDefinition.coinName+"-paymentack"; // BIP 71

	public static Protos.PaymentRequest createPaymentRequest(final BigInteger amount, @Nonnull final Address toAddress, final String memo,
			final String paymentUrl)
	{
		if (amount != null && amount.compareTo(BigInteger.valueOf(Long.MAX_VALUE)) > 0)
			throw new IllegalArgumentException("amount too big for protobuf: " + amount);

		final Protos.Output.Builder output = Protos.Output.newBuilder();
		output.setAmount(amount != null ? amount.longValue() : 0);
		output.setScript(ByteString.copyFrom(ScriptBuilder.createOutputScript(toAddress).getProgram()));

		final Protos.PaymentDetails.Builder paymentDetails = Protos.PaymentDetails.newBuilder();
		paymentDetails.setNetwork(Constants.NETWORK_PARAMETERS.getPaymentProtocolId());
		paymentDetails.addOutputs(output);
		if (memo != null)
			paymentDetails.setMemo(memo);
		if (paymentUrl != null)
			paymentDetails.setPaymentUrl(paymentUrl);
		paymentDetails.setTime(System.currentTimeMillis());

		final Protos.PaymentRequest.Builder paymentRequest = Protos.PaymentRequest.newBuilder();
		paymentRequest.setSerializedPaymentDetails(paymentDetails.build().toByteString());

		return paymentRequest.build();
	}

	public static PaymentIntent parsePaymentRequest(@Nonnull final byte[] serializedPaymentRequest) throws PaymentRequestException
	{
		try
		{
			if (serializedPaymentRequest.length > 50000)
				throw new PaymentRequestException("payment request too big: " + serializedPaymentRequest.length);

			final Protos.PaymentRequest paymentRequest = Protos.PaymentRequest.parseFrom(serializedPaymentRequest);

			final String pkiName;
			final String pkiOrgName;
			final String pkiCaName;
			if (!"none".equals(paymentRequest.getPkiType()))
			{
				// implicitly verify PKI signature
				final PkiVerificationData verificationData = new PaymentSession(paymentRequest, true).pkiVerificationData;
				pkiName = verificationData.name;
				pkiOrgName = verificationData.orgName;
				pkiCaName = verificationData.rootAuthorityName;
			}
			else
			{
				pkiName = null;
				pkiOrgName = null;
				pkiCaName = null;
			}

			if (paymentRequest.getPaymentDetailsVersion() != 1)
				throw new PaymentRequestException.InvalidVersion("cannot handle payment details version: "
						+ paymentRequest.getPaymentDetailsVersion());

			final Protos.PaymentDetails paymentDetails = Protos.PaymentDetails.newBuilder().mergeFrom(paymentRequest.getSerializedPaymentDetails())
					.build();

			final long currentTimeSecs = System.currentTimeMillis() / 1000;
			if (paymentDetails.hasExpires() && currentTimeSecs >= paymentDetails.getExpires())
				throw new PaymentRequestException.Expired("payment details expired: current time " + currentTimeSecs + " after expiry time "
						+ paymentDetails.getExpires());

			if (!paymentDetails.getNetwork().equals(Constants.NETWORK_PARAMETERS.getPaymentProtocolId()))
				throw new PaymentRequestException.InvalidNetwork("cannot handle payment request network: " + paymentDetails.getNetwork());

			final ArrayList<PaymentIntent.Output> outputs = new ArrayList<PaymentIntent.Output>(paymentDetails.getOutputsCount());
			for (final Protos.Output output : paymentDetails.getOutputsList())
				outputs.add(parseOutput(output));

			final String memo = paymentDetails.hasMemo() ? paymentDetails.getMemo() : null;
			final String paymentUrl = paymentDetails.hasPaymentUrl() ? paymentDetails.getPaymentUrl() : null;
			final byte[] merchantData = paymentDetails.hasMerchantData() ? paymentDetails.getMerchantData().toByteArray() : null;

			final PaymentIntent paymentIntent = new PaymentIntent(PaymentIntent.Standard.BIP70, pkiName, pkiOrgName, pkiCaName,
					outputs.toArray(new PaymentIntent.Output[0]), memo, paymentUrl, merchantData, null);

			if (paymentIntent.hasPaymentUrl() && !paymentIntent.isSupportedPaymentUrl())
				throw new PaymentRequestException.InvalidPaymentURL("cannot handle payment url: " + paymentIntent.paymentUrl);

			return paymentIntent;
		}
		catch (final InvalidProtocolBufferException x)
		{
			throw new PaymentRequestException(x);
		}
		catch (final UninitializedMessageException x)
		{
			throw new PaymentRequestException(x);
		}
	}

	private static PaymentIntent.Output parseOutput(@Nonnull final Protos.Output output) throws PaymentRequestException.InvalidOutputs
	{
		try
		{
			final BigInteger amount = BigInteger.valueOf(output.getAmount());
			final Script script = new Script(output.getScript().toByteArray());
			return new PaymentIntent.Output(amount, script);
		}
		catch (final ScriptException x)
		{
			throw new PaymentRequestException.InvalidOutputs("unparseable script in output: " + output.toString());
		}
	}

	public static Protos.Payment createPaymentMessage(@Nonnull final Transaction transaction, @Nullable final Address refundAddress,
			@Nullable final BigInteger refundAmount, @Nullable final String memo, @Nullable final byte[] merchantData)
	{
		final Protos.Payment.Builder builder = Protos.Payment.newBuilder();

		builder.addTransactions(ByteString.copyFrom(transaction.unsafeBitcoinSerialize()));

		if (refundAddress != null)
		{
			if (refundAmount.compareTo(BigInteger.valueOf(Long.MAX_VALUE)) > 0)
				throw new IllegalArgumentException("refund amount too big for protobuf: " + refundAmount);

			final Protos.Output.Builder refundOutput = Protos.Output.newBuilder();
			refundOutput.setAmount(refundAmount.longValue());
			refundOutput.setScript(ByteString.copyFrom(ScriptBuilder.createOutputScript(refundAddress).getProgram()));
			builder.addRefundTo(refundOutput);
		}

		if (memo != null)
			builder.setMemo(memo);

		if (merchantData != null)
			builder.setMerchantData(ByteString.copyFrom(merchantData));

		return builder.build();
	}

	public static List<Transaction> parsePaymentMessage(final Protos.Payment paymentMessage)
	{
		final List<Transaction> transactions = new ArrayList<Transaction>(paymentMessage.getTransactionsCount());

		for (final ByteString transaction : paymentMessage.getTransactionsList())
			transactions.add(new Transaction(Constants.NETWORK_PARAMETERS, transaction.toByteArray()));

		return transactions;
	}

	public static Protos.PaymentACK createPaymentAck(@Nonnull final Protos.Payment paymentMessage, @Nullable final String memo)
	{
		final Protos.PaymentACK.Builder builder = Protos.PaymentACK.newBuilder();

		builder.setPayment(paymentMessage);

		builder.setMemo(memo);

		return builder.build();
	}

	public static String parsePaymentAck(@Nonnull final Protos.PaymentACK paymentAck)
	{
		final String memo = paymentAck.hasMemo() ? paymentAck.getMemo() : null;

		return memo;
	}
}


wallet/src/de/schildbach/wallet/Constants.java
/*
 * Copyright 2011-2014 the original author or authors.
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

package de.schildbach.wallet;

import java.io.File;
import java.nio.charset.Charset;

import android.os.Build;
import android.os.Environment;
import android.text.format.DateUtils;

import com.google.bitcoin.core.NetworkParameters;
import com.google.bitcoin.params.MainNetParams;
//import com.google.bitcoin.params.TestNet3Params;
import com.google.bitcoin.core.CoinDefinition;
import hashengineering.darkcoin.wallet.R;

/**
 * @author Andreas Schildbach
 */
public class Constants
{
	public static final boolean TEST = false; //R.class.getPackage().getName().contains("_test");

	public static final NetworkParameters NETWORK_PARAMETERS = MainNetParams.get();
//	private static final String FILENAME_NETWORK_SUFFIX = NETWORK_PARAMETERS.getId().equals(NetworkParameters.ID_MAINNET) ? "" : "-testnet";


	public static final String WALLET_FILENAME_PROTOBUF = "wallet-protobuf";

	public static final String WALLET_KEY_BACKUP_BASE58 = "key-backup-base58";
	public static final String WALLET_KEY_BACKUP_PROTOBUF = "key-backup-protobuf";

	public static final File EXTERNAL_WALLET_BACKUP_DIR = Environment.getExternalStoragePublicDirectory(Environment.DIRECTORY_DOWNLOADS);
	public static final String EXTERNAL_WALLET_KEY_BACKUP = CoinDefinition.coinName +"-wallet-keys";

	public static final String BLOCKCHAIN_FILENAME = "blockchain";

	public static final String CHECKPOINTS_FILENAME = "checkpoints";

	private static final String EXPLORE_BASE_URL_PROD = CoinDefinition.BLOCKEXPLORER_BASE_URL_PROD;
//	private static final String EXPLORE_BASE_URL_TEST = CoinDefinition.BLOCKEXPLORER_BASE_URL_TEST;
//	public static final String EXPLORE_BASE_URL = NETWORK_PARAMETERS.getId().equals(NetworkParameters.ID_MAINNET) ? EXPLORE_BASE_URL_PROD
//			: EXPLORE_BASE_URL_TEST;
    public static final String EXPLORE_BASE_URL = EXPLORE_BASE_URL_PROD;
    public static final String EXPLORE_ADDRESS_PATH  = CoinDefinition.BLOCKEXPLORER_ADDRESS_PATH;
    public static final String EXPLORE_TRANSACTION_PATH  = CoinDefinition.BLOCKEXPLORER_TRANSACTION_PATH;
    public static final String EXPLORE_BLOCK_PATH  = CoinDefinition.BLOCKEXPLORER_BLOCK_PATH;


	public static final String MIMETYPE_PAYMENTREQUEST = "application/"+ CoinDefinition.coinName +"-paymentrequest"; // BIP 71
	public static final String MIMETYPE_PAYMENT = "application/"+ CoinDefinition.coinName +"-payment"; // BIP 71
	public static final String MIMETYPE_PAYMENTACK = "application/"+ CoinDefinition.coinName +"-paymentack"; // BIP 71
	public static final String MIMETYPE_TRANSACTION = "application/x-" + CoinDefinition.coinName + "tx";
	public static final String MIMETYPE_BACKUP_PRIVATE_KEYS = "x-"+CoinDefinition.coinName+"/private-keys";


	public static final int MAX_NUM_CONFIRMATIONS = 7;
	public static final String USER_AGENT = CoinDefinition.coinName +" Wallet";
	public static final String DEFAULT_EXCHANGE_CURRENCY = "USD";
	public static final int WALLET_OPERATION_STACK_SIZE = 256 * 1024;
	public static final long BLOCKCHAIN_STATE_BROADCAST_THROTTLE_MS = DateUtils.SECOND_IN_MILLIS;
	public static final long BLOCKCHAIN_UPTODATE_THRESHOLD_MS = DateUtils.HOUR_IN_MILLIS;

	public static final String CURRENCY_CODE_BTC = CoinDefinition.coinTicker;
	public static final String CURRENCY_CODE_MBTC = "m" + CoinDefinition.coinTicker;
	public static final String CURRENCY_CODE_UBTC = "µ" + CoinDefinition.coinTicker;

	public static final char CHAR_HAIR_SPACE = '\u200a';
	public static final char CHAR_THIN_SPACE = '\u2009';
	public static final char CHAR_ALMOST_EQUAL_TO = '\u2248';
	public static final char CHAR_CHECKMARK = '\u2713';
	public static final String CURRENCY_PLUS_SIGN = "+" + CHAR_THIN_SPACE;
	public static final String CURRENCY_MINUS_SIGN = "-" + CHAR_THIN_SPACE;
	public static final String PREFIX_ALMOST_EQUAL_TO = Character.toString(CHAR_ALMOST_EQUAL_TO) + CHAR_THIN_SPACE;
	public static final int ADDRESS_FORMAT_GROUP_SIZE = 4;
	public static final int ADDRESS_FORMAT_LINE_SIZE = 12;

	public static final int BTC_MAX_PRECISION = 8;
	public static final int MBTC_MAX_PRECISION = 5;

	public static final int LOCAL_PRECISION = 8;        //altcoins need more digits in BTC
    //TODO: What about these two things?
	public static final String DONATION_ADDRESS = CoinDefinition.DONATION_ADDRESS;
	public static final String REPORT_EMAIL = "setafur@gmail.com";

	public static final int UBTC_MAX_PRECISION = 2;


	public static final String REPORT_SUBJECT_ISSUE = "Reported issue";
	public static final String REPORT_SUBJECT_CRASH = "Crash report";

	public static final String LICENSE_URL = "http://www.gnu.org/licenses/gpl-3.0.txt";

    public static final String FORKED_FROM_SOURCE = "based on bitcoin-, darkcoin-, and peercoin-wallet\n";
    public static final String FORKED_FROM_SOURCE_BITCOINJ = "based on bitcoinj, darkcoinj, and peercoinj\n";
	public static final String SOURCE_URL = "https://github.com/stafur/" + CoinDefinition.coinName.toLowerCase() + "-wallet";
	public static final String BINARY_URL = "https://github.com/stafur/"+ CoinDefinition.coinName.toLowerCase() +"-wallet/releases";
	public static final String CREDITS_BITCOINJ_URL = "https://github.com/stafur/" + CoinDefinition.coinName.toLowerCase() + "coinj";
	public static final String CREDITS_ZXING_URL = "http://code.google.com/p/zxing/";
    public static final String CREDITS_WEBSITE_URL = "http://trustplus.co/";
    public static final String CREDITS_FORUM_URL = "https://darkcointalk.org/";

	public static final String CREDITS_ICON_URL = "https://bitcointalk.org/index.php?topic=687207.msg8973682#msg8973682";

	public static final String AUTHOR_TWITTER_URL = "https://twitter.com/TrustPlusCoin";
	public static final String AUTHOR_GOOGLEPLUS_URL = "https://profiles.google.com/setafur";

	public static final String COMMUNITY_GOOGLEPLUS_URL = "https://plus.google.com/setafur";

	public static final String MARKET_APP_URL = "market://details?id=%s";
	public static final String WEBMARKET_APP_URL = "https://play.google.com/store/apps/details?id=%s";
	public static final String MARKET_PUBLISHER_URL = "market://search?q=pub:\"Sergio Tafur\"";

	public static final String VERSION_URL = "http://wallet.schildbach.de/version";
	public static final int HTTP_TIMEOUT_MS = 15 * (int) DateUtils.SECOND_IN_MILLIS;

	public static final long LAST_USAGE_THRESHOLD_JUST_MS = DateUtils.HOUR_IN_MILLIS;
	public static final long LAST_USAGE_THRESHOLD_RECENTLY_MS = 2 * DateUtils.DAY_IN_MILLIS;

	public static final int SDK_JELLY_BEAN = 16;
	public static final int SDK_JELLY_BEAN_MR2 = 18;

	public static final int SDK_DEPRECATED_BELOW = Build.VERSION_CODES.ICE_CREAM_SANDWICH;

	public static final boolean BUG_OPENSSL_HEARTBLEED = Build.VERSION.SDK_INT == Constants.SDK_JELLY_BEAN
			&& Build.VERSION.RELEASE.startsWith("4.1.1");

	public static final int MEMORY_CLASS_LOWEND = 48;

	public static final Charset UTF_8 = Charset.forName("UTF-8");
	public static final Charset US_ASCII = Charset.forName("US-ASCII");
}
